Index: com/limegroup/gnutella/settings/iTunesSettings.java
===================================================================
RCS file: /usr/local/tigris/data/helm/cvs/repository/core/com/limegroup/gnutella/settings/iTunesSettings.java,v
retrieving revision 1.3
diff -u -r1.3 iTunesSettings.java
--- com/limegroup/gnutella/settings/iTunesSettings.java	2003/09/29 20:04:43	1.3
+++ com/limegroup/gnutella/settings/iTunesSettings.java	2004/02/21 23:54:48
@@ -22,4 +22,35 @@
         FACTORY.createStringArraySetting("ITUNES_SUPPORTED_FILE_TYPES", 
             new String[]{".mp3", ".aif", ".aiff", ".wav", ".mp2", ".mp4", 
                         ".aac", ".mid", ".m4a", ".m4p", ".ogg"});
+						
+	
+	// DAAP
+	
+	public static BooleanSetting DAAP_SUPPORT_ENABLED =
+		FACTORY.createBooleanSetting("DAAP_SUPPORT_ENABLED", true);
+		
+	public static StringSetting DAAP_LIBRARY_NAME =
+		FACTORY.createStringSetting("DAAP_LIBRARY_NAME", "LimeWire");
+	
+	public static IntSetting DAAP_PORT =
+		FACTORY.createIntSetting("DAAP_PORT", 5214);
+		
+	public static StringSetting DAAP_TYPE_NAME =
+		FACTORY.createStringSetting("DAAP_TYPE_NAME", "_daap._tcp.local.");
+	
+	//  Actually: "LimeWire" + "." + DAAP_TYPE_NAME.getValue()
+	public static StringSetting DAAP_SERVICE_NAME =
+		FACTORY.createStringSetting("DAAP_SERVICE_NAME", "LimeWire");
+		
+	public static IntSetting DAAP_WEIGHT = FACTORY.createIntSetting("DAAP_WEIGHT", 0);
+	public static IntSetting DAAP_PRIORITY = FACTORY.createIntSetting("DAAP_PRIORITY", 0);
+	
+	public static StringSetting DAAP_SERVICE_DESCRIPTION =
+		FACTORY.createStringSetting("DAAP_SERVICE_DESCRIPTION", "LimeWire");
+		
+	public static BooleanSetting DAAP_REQUIRES_PASSWORD =
+		FACTORY.createBooleanSetting("DAAP_REQUIRES_PASSWORD", false);
+		
+	public static StringSetting DAAP_PASSWORD =
+		FACTORY.createStringSetting("DAAP_PASSWORD", "");
 }
